### ISSUE TITLE
Enabling Actor Partitioning caused hang in registering reminder with Postgres

### ISSUE THREAD

**Issue by voloda** [dapr/dapr#3533]
## In what area(s)?

/area runtime

> /area operator

> /area placement

> /area docs

> /area test-and-release

## What version of Dapr?

1.3.0

## Expected Behavior

Registration quickly returns with success

## Actual Behavior

Enabling the experimental feature causes the API to hang

## Steps to Reproduce the Problem

* Enable the experimental feature for Actor reminder partitioning in config:
```
  features:
    - name: Actor.TypeMetadata
      enabled: true
```
* Setup the actor state storage to PGSQL 13 (not sure if version matters)
* Set the number of partitions to any number (4, 10, ...)
* Try to register a reminder from within an executing actor
* I was trying to enable partitioning to see how the behavior changes under load due
* I also re-created PGSQL database just to be sure there are not some bad data
* Some additional background:
  * GKE hosting with HA
  * Deployed with HELM
  * Running backend services written in .NET 5
  * Setting the experimental flag to `false` solves the issue
* I was trying to look for some details in Stackdriver but there is no special evidence except the HttpClient timeout upon the actor invocation

## Release Note

RELEASE NOTE: FIX Postgres support for Actors.

**Comment by voloda** [comment]
> @voloda does this still reproduce with a different state store provider?

With REDIS this seems to work just fine (have REDIS deployed to same GKE using the default setup in Bitnami HELM chart) - e.g. 3 replicas

**Comment by artursouza** [comment]
The next steps for this issue is to add Postgres to conformance tests in contrib to make sure it behaves as expected for eTag matching operations. Because I worked on the reminder partition problem, my guess is that this issue will be reflected in the conformance tests for the component.

**Comment by beiwei30** [comment]
@artursouza I tried to add state conformance test for Postgres with dapr/components-contrib#1118, and got the following result:

```
--- FAIL: TestStateConformance (0.57s)
    --- FAIL: TestStateConformance/postgresql (0.57s)
        --- PASS: TestStateConformance/postgresql/init (0.04s)
        --- PASS: TestStateConformance/postgresql/ping (0.00s)
        --- FAIL: TestStateConformance/postgresql/set (0.17s)
        --- FAIL: TestStateConformance/postgresql/get (0.02s)
        --- FAIL: TestStateConformance/postgresql/delete (0.04s)
        --- FAIL: TestStateConformance/postgresql/bulkset (0.10s)
        --- FAIL: TestStateConformance/postgresql/bulkdelete (0.02s)
        --- FAIL: TestStateConformance/postgresql/transaction (0.15s)
        --- FAIL: TestStateConformance/postgresql/etag (0.01s)
        --- FAIL: TestStateConformance/postgresql/first-write_without_etag (0.01s)
        --- FAIL: TestStateConformance/postgresql/first-write_with_etag (0.01s)
FAIL
FAIL	github.com/dapr/components-contrib/tests/conformance	1.157s
FAIL
```

Most of tests fail due to the following error:

```
    state.go:584: 
        	Error Trace:	state.go:584
        	Error:      	Expected nil, but got: &pgconn.PgError{Severity:"ERROR", Code:"22P02", Message:"invalid input syntax for type json", Detail:"Token \"testValue2\" is invalid.", Hint:"", Position:0, InternalPosition:0, InternalQuery:"", Where:"JSON data, line 1: testValue2", SchemaName:"", TableName:"", ColumnName:"", DataTypeName:"", ConstraintName:"", File:"jsonfuncs.c", Line:621, Routine:"json_ereport_error"
```

**Comment by artursouza** [comment]
@beiwei30 Thanks a lot. The component will need to be fixed then, maybe rewritten.

**Comment by beiwei30** [comment]
@artursouza I try to make postgresql to pass the test, pls. check  https://github.com/dapr/components-contrib/pull/1118#issuecomment-912381915

**Comment by CodeMonkeyLeet** [comment]
> @artursouza I try to make postgresql to pass the test, pls. check [dapr/components-contrib#1118 (comment)](https://github.com/dapr/components-contrib/pull/1118#issuecomment-912381915)

@artursouza Is there any reason that the postgresql component needs/wants to define the state store table as type `json` and not `text`? It doesn't seem to take advantage of that column type in the implementation anyway, and the use of `text` simplifies the fact that the conformance tests expect binary blobs are retrieved unmodified on `Get()`, which JSON would require to be base64-encoded (and not just `\x` character encoded). 

On a related note, the conformance tests enforce that state components will marshal any type that is not a raw `[]byte` to JSON string on `Set()`, and return the raw `[]byte` representation of marshalled JSON (as suggested by the [Get State API](https://docs.dapr.io/reference/api/state_api/#get-state)) but it doesn't talk about the expectation that `[]byte` inputs have his special handling. Is that a docs gap? I'm guessing it's to support use cases where the app performs its own serialization/deserialization to binary but that's not really discussed anywhere. 

From a representation in the database perspective, depending on the implementation, `[]byte` arrays can be straight up stringified arrays (i.e. character escape strings) or base64-encoded, but all nested arrays or array members in structs will usually be base64-encoded because of the JSON marshalling. Does Dapr care to provide guidance on this one way or the other (i.e. for efficiency/consistency reasons)?

**Comment by artursouza** [comment]
> > @artursouza I try to make postgresql to pass the test, pls. check [dapr/components-contrib#1118 (comment)](https://github.com/dapr/components-contrib/pull/1118#issuecomment-912381915)
> 
> @artursouza Is there any reason that the postgresql component needs/wants to define the state store table as type `json` and not `text`? It doesn't seem to take advantage of that column type in the implementation anyway, and the use of `text` simplifies the fact that the conformance tests expect binary blobs are retrieved unmodified on `Get()`, which JSON would require to be base64-encoded (and not just `\x` character encoded).
> 
> On a related note, the conformance tests enforce that state components will marshal any type that is not a raw `[]byte` to JSON string on `Set()`, and return the raw `[]byte` representation of marshalled JSON (as suggested by the [Get State API](https://docs.dapr.io/reference/api/state_api/#get-state)) but it doesn't talk about the expectation that `[]byte` inputs have his special handling. Is that a docs gap? I'm guessing it's to support use cases where the app performs its own serialization/deserialization to binary but that's not really discussed anywhere.
> 
> From a representation in the database perspective, depending on the implementation, `[]byte` arrays can be straight up stringified arrays (i.e. character escape strings) or base64-encoded, but all nested arrays or array members in structs will usually be base64-encoded because of the JSON marshalling. Does Dapr care to provide guidance on this one way or the other (i.e. for efficiency/consistency reasons)?


There are two aspects here. First, is that the components API does not have a content-type header and it handles values as interface{}. Ideally, the value should be byte[] with content-type specifying which value it is. Then the database component can decide how to handle based on the type.

The second thing is how each component choses to save data. CosmosDB has special logic to parse binary data but the lack of content-type creates a problem too. I think each component must understand how to handle the special case for byte[] and I think the conformance test should cover the generic behavior and the test plan for the integration tests under discussion should handle testing for components to validate how efficiently they are saving data.

**Comment by artursouza** [comment]
Issue is still going on. I could reproduce using the E2E test with postgres as state store. The following error is printed in the daprd container of the test app:

```txt
time="2021-11-18T00:40:30.929048137Z" level=error msg="error getting reminders for actor type testactorreminderpartition: could not read actor type metadata: invalid etag value: strconv.Atoi: parsing \"\": invalid syntax" app_id=actorreminderpartition instance=actorreminderpartition-7977f574fb-trs5x scope=dapr.runtime.actor type=log ver=edge
```

To repro:
```txt
make delete-minikube; make setup-minikube && \
`make describe-minikube-env` && \
make build-linux && \
make docker-build && \
make docker-push && \
make create-test-namespace && \
make setup-test-env-redis && \
make setup-test-env-kafka && \
make docker-deploy-k8s && \
make setup-app-configurations && \
make setup-test-components && \
make build-e2e-app-all && \
make push-e2e-app-all

helm repo add bitnami https://charts.bitnami.com/bitnami
helm install postgresql -n dapr-tests bitnami/postgresql
# print password
kubectl get secret --namespace dapr-tests postgresql -o jsonpath="{.data.postgresql-password}" | base64 --decode
```
```txt
kubectl apply -f -

apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: statestore
  namespace: dapr-tests
spec:
  type: state.postgresql
  version: v1
  initTimeout: 1m
  metadata:
  - name: connectionString
    value: "host=postgresql.dapr-tests.svc.cluster.local user=postgres password=[paste password here] port=5432 connect_timeout=10 database=postgres"
  - name: actorStateStore
    value: true
```

```txt
DAPR_E2E_TEST=actor_reminder_partition make test-e2e-all
```

**Comment by DeepanshuA** [comment]
I further verified, this issue seems to be fixed by https://github.com/dapr/components-contrib/pull/1498/commits/e4ad31d36cc28fd3b0c0c97e6d416ed24819e546.
I reverted this commit and got an issue while running actor-reminder-partition tests. Though, it is not same as what Artur received as per previous comment, but it fails. Seems record is not getting saved properly, so while going to delete, it is failing. Also, record in postgres doesn't seem good.


tests/e2e/actor_reminder_partition TestActorReminder/Actor_reminder_changes_number_of_partitions. (15.00s)
    actor_reminder_partition_test.go:195:
                Error Trace:    actor_reminder_partition_test.go:195
                Error:          Received unexpected error:
                                Delete "http://192.168.49.2:30188/test/testactorreminderpartition/actor-reminder-partition-test-1000/reminders/PartitionTestReminder": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
                Test:           TestActorReminder/Actor_reminder_changes_number_of_partitions.
    --- FAIL: TestActorReminder/Actor_reminder_changes_number_of_partitions. (15.00s)

=== FAIL: tests/e2e/actor_reminder_partition TestActorReminder (16.01s)
2022/03/31 11:12:35 Waiting until service ingress is ready for actorreminderpartition...
2022/03/31 11:12:35 Service ingress for actorreminderpartition is ready...
    actor_reminder_partition_test.go:176: Checking if app is healthy ...

DONE 2 tests, 2 failures in 43.720s
make: *** [tests/dapr_tests.mk:220: test-e2e-all] Error 1



postgres=# select value from public.state;
                                              value
-------------------------------------------------------------------------------------------------
 null
 {"id": "00000000-0000-0000-0000-000000000000", "actorRemindersMetadata": {"partitionCount": 0}}
(2 rows)

**Comment by artursouza** [comment]
Closing issue since @DeepanshuA confirmed it has been fixed in 1.7
